astropy/io/fits/header.py

```diff
@@ -2010,8 +2011,20 @@ def fromfile(cls, fileobj):
         done with the parse_header function implemented in Cython."""
 
         close_file = False
-        if isinstance(fileobj, str):
-            fileobj = open(fileobj, 'rb')
+
+        if isinstance(fileobj, path_like):
+            # If sep is non-empty we are trying to read a header printed to a
+            # text file, so open in text mode by default to support newline
+            # handling; if a binary-mode file object is passed in, the user is
+            # then on their own w.r.t. newline handling.
+            #
+            # Otherwise assume we are reading from an actual FITS file and open
+            # in binary mode.
+            if sep:
+                fileobj = open(fileobj, 'r', encoding='latin1')
+            else:
+                fileobj = open(fileobj, 'rb')
+
```

Comment:
This block should be moved to the `Header.fromfile` method above. Here the method should always use bytes.